design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+
+
+## Public APIS
+
+### Test File
+
+The test file is a JSON file that describes the test settings and test data. The following is an example test file:
+
+```JSON
+{
+ "comments": [
+   "Any comments on this test file!"
+ ],
+ "tests": [
+   {
+     "name": "Sample test",
+     "statements": [
+       "CREATE STREAM TEST (ID bigint, NAME varchar, VALUE double) WITH (kafka_topic='test_topic', value_format='DELIMITED', key='ID');",
+       "CREATE STREAM S1 as SELECT name FROM test where id > 100;"
+     ],
+     "inputs": [
+       {"topic": "test_topic", "key": 0, "value": "0,zero,0.0"},
+       {"topic": "test_topic", "key": 100, "value": "100,100,0.0"},
+       {"topic": "test_topic", "key": 101, "value": "101,101,0.0"}
+     ],
+     "outputs": [
+       {"topic": "S1", "key": 101, "value": "101"}
+     ]
+   }
+ ]
+}
+```
+
+Each test file can include multiple tests. Each test can have the following attributes (these descriptions are the same as the ones described ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md))):
+
+- _name_ (required): The name of the test case.
+description (optional): A description of what the test case is testing.
+- _format_ (optional): An array of multiple different formats to run the test case as, e.g. AVRO, JSON, DELIMITED.
+- _statements_ (required): The list of statements to execute as this test case.
+- _properties_ (optional): A map of property name to value. Can contain any valid Ksql config. The config is passed to the engine when executing the statements in the test case
+- _topics_ (optional): An array of the topics this test case needs. Allows more information about the topic to be supplied, e.g. an existing Avro schema.
+- _inputs_ (required if `expectedException` not supplied): The set of input messages to be produced to Kafka topic(s).
+- _outputs_ (required if `expectedException` not supplied) The set of output messages expected in the output topic(s).
+- _expectedException_ (required if `inputs` and `outputs` not supplied) The exception that should be thrown when executing the supplied statements.
+
+
+For more details on the test file format click ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md)).
+
+
+### Testing Tool
+
+The testing tool will be a command line tool called ksql-testing-tool that accepts a test file as input parameter. The following is the way we can use the tool:
+
+```shell
+$ ksql-testing-tool /path/to/the/testfile.json
+```
+The tool will run the tests from the test file and if all the tests pass will terminate successfully printing the success message in the terminal! If any of the tests fail the tool will print out the reason for the failure in the terminal. The following is an example of successful test:
+
+```shell
+$ ksql-testing-tool /path/to/passing_test.json
+All tests passed!
```

Comment:
We should also not just say "All tests passed", because it doesn't tell you whether even a single test was actually run.  If we want to include a summary line (which we should), then it should be like:

```
All 36 tests passed!
```
or, inspired by `mvn test`:
```
Tests run: 36, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.004 s
```